Add user input for DX cooling coil low ambient temp cutout

rraustad opened the pull request:
Pull request overview
---------------------
DX heating coils are typically controlled off during low ambient conditions and have that control in EnergyPlus. DX cooling coils may also include controls that disable operation of the compressor at low ambient temperatures. This new input allows users to control this aspect of DX cooling coil operation. Issue #6149 is addressed with this change.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [x] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rraustad committed:
Initial changes to DX coils for low ambient temp cutout

rraustad commented on doc/input-output-reference/src/overview/group-heating-and-cooling-coils.tex:
```diff
@@ -1825,6 +1825,10 @@ \subsubsection{Inputs}\label{inputs-13-004}
+\paragraph{Field: Minimum Outdoor Dry-Bulb Temperature for Compressor Operation}\label{field-minimum-outdoor-dry-bulb-temperature-for-compressor-operation}
+
+This numeric field defines the minimum outdoor air dry-bulb temperature where the cooling coil compressor turns off. If this input field is left blank, the default value is -25 C.
+
```
...input field is left blank, the value defaults to that specified in the IDD ??

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -7388,7 +7391,7 @@ namespace DXCoils {
 		DXCoil( DXCoilNum ).PrintLowOutTempMessage = false;
 
-		if ( ( AirMassFlow > 0.0 ) && ( GetCurrentScheduleValue( DXCoil( DXCoilNum ).SchedPtr ) > 0.0 || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterPumped || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterWrapped ) && ( PartLoadRatio > 0.0 ) && ( CompOp == On ) ) { // for cycling fan, reset mass flow to full on rate
+		if ( ( AirMassFlow > 0.0 ) && ( GetCurrentScheduleValue( DXCoil( DXCoilNum ).SchedPtr ) > 0.0 || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterPumped || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterWrapped ) && ( PartLoadRatio > 0.0 ) && ( CompOp == On ) && CompAmbTemp > DXCoil( DXCoilNum ).MinOATCompressor) { // criteria for coil operation
```
hmmm, so there's no avail schedule for HPWH's and they just run. No real problem here except that there are many times when the HPWH should be off and it passes through the ON code and does unnecessary calculations. Init can't figure out if it should run and set AirMassFlow == 0 ?

    && ( GetCurrentScheduleValue( DXCoil( DXCoilNum ).SchedPtr ) > 0.0 || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterPumped || DXCoil( DXCoilNum ).DXCoilType_Num == CoilDX_HeatPumpWaterHeaterWrapped ) &&

mjwitte replied:
Here we're testing CompAmbTemp but later it's CondInletTemp which seems more appropriate.  Is CompAmbTemp correct here?  Given that the condenser air inlet node isn't necessarily an outdoor air node, should the new field name be "Minimum Condenser Inlet Node Dry-Bulb Temperature for Compressor Operation"?

rraustad commented:
Revised unit tests and example files based on CI results. Not sure I know what to do for HVACTemplate files but they still need updates.

rraustad committed:
Update example files and unit tests

rraustad commented:
@mjwitte  pretty much step 1 done for Coil:Cooling:DX:SingleSpeed. Comments on implementation before progressing? I'm going to hold until I hear back.

rraustad commented on testfiles/SupermarketTwoStageFlashIntercooler.idf:
```diff
@@ -1515,7 +1516,7 @@
     Coil:Heating:Fuel,       !- Heating Coil Object Type
     Furnace Heating Coil 1,  !- Heating Coil Name
-    Coil:Cooling:DX:SingleSpeed,  !- Cooling Coil Object Type
+    Coil:Cooling:\ingleSpeed,  !- Cooling Coil Object Type
```
Oops

rraustad commented:
I'm a little confused as to the elimination of these errors. If default is -25C, then shouldn't these still report?

    -   ** Warning ** CalcDoe2DXCoil: Coil:Cooling:DX:SingleSpeed "PSZ-AC:1_COOLC DXCOIL" - Air-cooled condenser inlet dry-bulb temperature below 0 C. Outdoor dry-bulb temperature = -20.60
    -   **   ~~~   **  ... Occurrence info = CHICAGO ANN HTG 99.6% CONDNS DB, 01/21 00:02 - 00:04
    -   **   ~~~   ** ... Operation at low ambient temperatures may require special performance curves.

Ah, this field is after min fields and if not included in input then will default to 0. Min fields need not be increased just for this field to default correctly, will handle in GetInput.

rraustad committed:
Update EPFilter and default min OAT if not included in input

rraustad committed:
2nd attempt at EPFilter changes, 6 other places modified.

rraustad commented:
This looks like its working correctly. This is UntarySystem with 23C cutout on cooling.

![minoat](https://cloud.githubusercontent.com/assets/5778156/26758346/6aced0fc-48a9-11e7-867f-2e1b77e8b418.png)

rraustad committed:
Add two-speed, multi-speed and variable-speed cooling coils.

rraustad commented:
Noticed that this didn't get updated when new coil types were added:

    AirLoopHVAC:Unitary:Furnace:HeatCool,
     A12, \field Cooling Coil Object Type
        \required-field
        \type choice
        \key Coil:Cooling:DX:SingleSpeed
        \key CoilSystem:Cooling:DX:HeatExchangerAssisted
        \note  Only works with DX cooling coil types

Code shows this. So it looks like a VS coil or integrated coil system HP are allowed. I only see VS coil in IO Ref for heat cool furnace.

    } else if ( Furnace( FurnaceNum ).CoolingCoilType_Num == Coil_CoolingAirToAirVariableSpeed ) {
        // BOS ADDED, AUG/2012, VARIIABLE SPEED DX COOLING COIL
        //  Furnace(FurnaceNum)%DXCoolCoilType = 'COIL:COOLING:DX:VARIABLESPEED'
        //  Furnace(FurnaceNum)%DXCoolCoilName = CoolingCoilName
        if ( SameString( CoolingCoilType, "COILSYSTEM:INTEGRATEDHEATPUMP:AIRSOURCE" ) ) Furnace( FurnaceNum ).bIsIHP = true;

rraustad committed:
Added multimode and multispeed DX cooling coils. Updated example files. Corrected unit test. Delete unused variable.

rraustad committed:
Some files weren't saved.

rraustad committed:
Correct multispeed coil GetInput

rraustad committed:
Fixed more mistakes. About done here.

rraustad commented:
With the above commit I'm in testing mode now. I've looked at Furnace, Heat Pump, and UnitarySystem.

![furnacewithdxsystemrhcontrol](https://user-images.githubusercontent.com/5778156/26842507-14b00d30-4abb-11e7-8cf2-f2a46db2d6f8.png)

![heatpumpairtoairrhcontrol](https://user-images.githubusercontent.com/5778156/26842513-1923e814-4abb-11e7-8b3a-0079944db9d3.png)

![unitarysystem_heatpumpauto](https://user-images.githubusercontent.com/5778156/26842521-1c9ca170-4abb-11e7-9f13-1fec462813fa.png)

rraustad commented:
PTHP's:

![packagedterminalheatpump](https://user-images.githubusercontent.com/5778156/26844441-6fa8e85a-4ac1-11e7-9f81-a6aa1cfbbc48.png)

![packagedterminalheatpumpvsas](https://user-images.githubusercontent.com/5778156/26844449-74126420-4ac1-11e7-99d8-f67e55bcfc12.png)

mjwitte commented on src/EnergyPlus/Furnaces.hh:
```diff
@@ -676,6 +678,18 @@ namespace Furnaces {
 	);
 
+	void
+	SetMinOATCompressor(
```
@rraustad Why go through all this effort to pass the min OAT up to the parent object and check it there?  Why not just check it in the coil object and let it return with zero load, same as if the coil schedule was off for that timestep?

rraustad replied:
I am checking min OAT in the coil objects. But I also wanted to see if I could eliminate unnecessary calculations when I know the coil would be off (e.g., if off return, or something like that). Turns out it was harder than I thought (for example I thought I could set the load = 0 in Init and be done but that also turns off the supp heater). There are existing places in the parent that mine that data and do the check so this replicates that method.

rraustad commented:
Easy comparison of new field placement.

![dxsinglenewminoat](https://user-images.githubusercontent.com/5778156/26846122-c958bab0-4ac6-11e7-8c04-331f9cbb0fca.png)

![dxmultiminoat](https://user-images.githubusercontent.com/5778156/26846136-cedca2bc-4ac6-11e7-953d-001eee50adb5.png)

![dxvariableminoat](https://user-images.githubusercontent.com/5778156/26846144-d274d4b2-4ac6-11e7-968d-97e312b3d29a.png)

![dxtwospeedminoat](https://user-images.githubusercontent.com/5778156/26846149-d677de7e-4ac6-11e7-88e0-8b59e4036c9c.png)

![dxtwostageminoat](https://user-images.githubusercontent.com/5778156/26846163-d9a7558e-4ac6-11e7-8777-675b80d726f9.png)

rraustad committed:
Correct HP compressor report variable. Depricate PTHP Min OAT since it's redundant. Clean up.

rraustad commented:
Still need transition rules.

rraustad committed:
Fix furnace air-to-air HP, PTHP IDD max field needed changing.

rraustad commented:
@Myoldmopar what do we do for transition changes? Is the new doc markdown or excel?

rraustad commented:
Never mind, I used md for #6108 and I'll just add to that doc.

rraustad commented:
Diff's in HeatPumpWaterToAirEquationFit, although the diff's are a correction to the compressor part-load ratio report variable. I believe the MinOATCompressor variable was 0 since these coils do not have that in;put. So when OAT was below 0 C, the compressor report variable was set to 0. Now MinOATCompressor = -1000.0 for these coils and the compressor part-load ratio report is accurate. However, this logic is flawed in that the heating coil was actually on so this code won't work for the Air-To-Air version.

![hpwattoaireqfit](https://user-images.githubusercontent.com/5778156/26907568-694d9d14-4bc2-11e7-9fac-f1cf01f24e62.png)

rraustad commented:
Diff's in VSHeatPumpWaterToAirEquationFit are reported as large when I can't see any real difference in results.

![vsheatpumpequationfit](https://user-images.githubusercontent.com/5778156/26946122-2e5bb110-4c5c-11e7-8e53-28bcf1421029.png)

rraustad committed:
Last commit? Transisition rules update. Unit test updates. Caught a few corrections.

rraustad commented:
I think I have explained all diffs with this one. All but 2 are AUD and Table Big Diffs given a new field this is expected. For the VS coil, I think I have explained that one as well (or missed the actual diff).

![auddiffs](https://user-images.githubusercontent.com/5778156/26958816-b8659c02-4c9a-11e7-8cdb-b1b9c8afe564.png)

rraustad commented:
I'm calling this good for now.

nrel-bot-2 commented:
@rraustad @lgentile it has been 15 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

mjwitte commented on doc/input-output-reference/src/overview/group-heating-and-cooling-coils.tex:
```diff
@@ -1825,6 +1825,10 @@ \subsubsection{Inputs}\label{inputs-13-004}
+\paragraph{Field: Minimum Outdoor Dry-Bulb Temperature for Compressor Operation}\label{field-minimum-outdoor-dry-bulb-temperature-for-compressor-operation}
+
+This numeric field defines the minimum outdoor air dry-bulb temperature where the cooling coil compressor turns off. If this input field is left blank, the value defaults to that specified in the IDD (e.g., -25 °C). If this field is not included in the input, the default value is -25 °C.
+
```
Seems a little redundant.  Should be adequate to say "If this field is left blank, it defaults to -25C."

rraustad replied:
Done

mjwitte commented on doc/input-output-reference/src/overview/group-heating-and-cooling-coils.tex:
```diff
@@ -3726,7 +3746,7 @@ \subsubsection{Inputs}\label{inputs-19-001}
 
-This numeric field defines the minimum outdoor air dry-bulb temperature where the heating coil compressor turns off. The temperature for this input field must be greater than or equal to 20 C. If this input field is left blank, the default value is -8 C.
+This numeric field defines the minimum outdoor air dry-bulb temperature where the heating coil compressor turns off. The temperature for this input field must be greater than or equal to 20 °C. If this input field is left blank, the default value is -8 °C.
 
```
Something isn't right here. ". . . must be greater than or equal to 20C.  . . . the default value is -8C."

rraustad replied:
should have been -20 C

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -43304,7 +43297,11 @@ Coil:Cooling:DX:SingleSpeed,
        \note cubic curve = a + b*PLR + c*PLR**2 + d*PLR**3
        \note PLR = part load ratio (cooling load/steady state capacity)
-   N6, \field Nominal Time for Condensate Removal to Begin
+  N6 , \field Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}
```
{C} should not be in the field name.  Same for all occurrences of this new field name, and where it was added to the existing field name in the dx heating coils.

rraustad replied:
Done

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -46210,7 +46223,7 @@ Coil:Heating:DX:SingleSpeed,
+  N5 , \field Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}
        \type real
        \minimum -20.0
        \default -8.0
```
3 DX heating coils, all have a field for MinOAT which defaults to -8.  But one has a minimum of -20, the next has no minimum, and the other has a minimum of -50.  Shouldn't these all be consistent and have no minimum?

rraustad replied:
For heating, I guess so.  But why a min for any of these coil types? Cooling could go down to -25C?

rraustad replied:
Minimums removed from all 3 DX heating coils

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -33653,13 +33653,6 @@ ZoneHVAC:PackagedTerminalHeatPump,
-        \default -8.0
-        \units C
-        \note Needs to match the corresponding minimum outdoor temperature defined
-        \note in the DX Heating Coil object.
```
Ok with removing this field from ZoneHVAC:PTHP.  What about the same field in AirLoopHVAC:UnitaryHeatPump:AirToAir:MultiSpeed?

rraustad replied:
I failed to see this one when implementing this. I guess it needs to go in.

rraustad replied:
It means more example file changes, more unit test changes and changing the code. Can I pick this one up with the next issue of the parent using the same OA condition as the child? Or do it now?

mjwitte replied:
At this point, later is fine.  I'll get started on the transition code.

Myoldmopar replied:
@mjwitte I'm happy to work up the transition code here if you'd like me to.

mjwitte replied:
@Myoldmopar That would be great.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -45192,7 +45201,11 @@ Coil:Cooling:DX:TwoStageWithHumidityControlMode,
        \type object-list
        \object-list WaterStorageTankNames
-  N5,  \field Basin Heater Capacity
+  N5,  \field Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}
```
Hmm, should this be in the CoilPerformance:DX:Cooling object instead of here?  The Condenser Type and Condenser Inlet Node are in CoilPerformance:DX:Cooling, but Maximum Outdoor Dry-Bulb Temperature for Crankcase Heater Operation is here in the master coil object.  (Who designed this beast anyway?)

rraustad replied:
Different OAT cutout for different stages? It's here because there would only be one minimum temp. I guess it's possible to have a different cutout for different stages.

mjwitte replied:
Maybe it's best to just leave it here as one value.

mjwitte commented on src/EnergyPlus/Furnaces.cc:
```diff
@@ -6274,36 +6287,38 @@ namespace Furnaces {
 						//       Check if Heat Pump compressor is allowed to run based on outdoor temperature
 						if ( Furnace( FurnaceNum ).CondenserNodeNum > 0 ) {
-							if ( Node( Furnace( FurnaceNum ).CondenserNodeNum ).Temp > Furnace( FurnaceNum ).MinOATCompressor ) {
+							if ( Node( Furnace( FurnaceNum ).CondenserNodeNum ).Temp > Furnace( FurnaceNum ).MinOATCompressorHeating ) {
```
Here, the test is on the condenser node temp.  OK.

rraustad replied:
CompAmbTemp = CondInletTemp, where OA conditions are set up at top of function

mjwitte commented on src/EnergyPlus/Furnaces.cc:
```diff
@@ -6274,36 +6287,38 @@ namespace Furnaces {
 							}
 						} else {
-							if ( OutdoorDryBulbTemp > Furnace( FurnaceNum ).MinOATCompressor ) {
+							if ( OutdoorDryBulbTemp > Furnace( FurnaceNum ).MinOATCompressorHeating ) {
```
But here, the test is on OutdoorDryBulbTemp?  This may need a separate issue to clean up this whole outdoor vs condenser/evaporator inlet node stuff.  In DXCoils, the variable naming is confusing with OutdoorDryBulb used to be the condenser inlet temp (even if it's not really the outdoor dry bulb) vs  OutDryBulbTemp (which *is* ODB).

rraustad replied:
In Furnace, OutdoorDryBulbTemp is either outdoor temp or the secondary zone temp. There is no direct input for condenser node name. This would have to be added.

rraustad replied:
OK, SetMinOATCompressor was added that mines the data from the child. It just wasn't set up to look at "node" conditions instead of OA condition.

rraustad replied:
Also only know MinOAT, not condenser node number.

mjwitte replied:
As I mentioned elsewhere - we need to clean this up so that parent and coil are looking at the same temps.  No reason the furnace can't ask for the coil condenser inlet nodenum and store that.  Not suggesting that be done as part of this.

mjwitte commented on src/EnergyPlus/Furnaces.cc:
```diff
@@ -9935,6 +9950,74 @@ namespace Furnaces {
+		std::string const CoolingCoilType, // type of cooling coil
+		std::string const CoolingCoilName, // name of cooling coil
+		std::string const HeatingCoilType, // type of heating coil
+		std::string const HeatingCoilName, // name of heating coil
```
Why use CoilName instead of CoilNum?  Don't we know the coil index yet when this is called?

rraustad replied:
Coils might not have been called yet when reading input for furnace. Could rewrite code to use coil index instead. Too much extra work.

rraustad replied:
Well, the coils have already been read in by the time this new function was called (at end of each loop). So yes, the coil index could have been passed instead of coil name.

Myoldmopar replied:
Yeah, I'm concerned about the performance hit due to this passing strings around.  If this was just being called once, it wouldn't be a big deal, but (and correct me if I'm wrong) it looks like this is the main `sim` interface, so this will be called lots of times.  If there is some reason strings should be passed here, then I think we are going to want to pass them via const reference  (`const std::string & CoolingCoilName`) to avoid making string copies.

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -8518,7 +8539,7 @@ namespace HVACUnitarySystem {
 
-			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressor ) {
+			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressorCooling) {
 				SimDXCoilMultiSpeed( CompName, UnitarySystem( UnitarySysNum ).CoolingSpeedRatio, UnitarySystem( UnitarySysNum ).CoolingCycRatio, CompIndex, UnitarySystem( UnitarySysNum ).CoolingSpeedNum, UnitarySystem( UnitarySysNum ).FanOpMode, CompOn, UnitarySystem( UnitarySysNum ).SingleMode );
```
Similar here - really should followup with a variable name change that's CondenserInletTemp instead of OutsideDryBulbTemp as a separate issue.  Same for PTHP/PTAC.

rraustad replied:
OK to new issue. Only remaining item is new input field name. There are a lot of places that would need to be changed.

mjwitte reviewed (commented):
@rraustad This branch has some conflicts, probably minor.  Once we settle on the new field name (see comments below) then I'll add the transition code.  Haven't run any tests on this yet.

rraustad committed:
Update after comments and merge develop.

Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #6149-Add-low-temperature-shut-off-to-air-cooled-DX-cooling-coils

# Conflicts:
#	src/EnergyPlus/HVACUnitarySystem.cc
#	src/Transition/InputRulesFiles/Rules8-7-0-to-8-8-0.md

rraustad commented on doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:
```diff
@@ -2651,7 +2647,6 @@ \subsubsection{Inputs}\label{inputs-9-026}
   Coil:Heating:DX:SingleSpeed,  !- Heating Coil Object Type
   Zone2PTHPDXHeatCoil,     !- Heating Coil Name
   0.001,                   !- Heating Convergence Tolerance {dimensionless}
-  2.0,                     !- Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}
```
Missed this one.

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -1895,9 +1895,9 @@ namespace HVACUnitarySystem {
 				PrimaryAirSystem( CurSysNum ).supFanModelTypeEnum = DataAirSystems::objectVectorOOFanSystemModel;
 			} else {
-				PrimaryAirSystem( CurSysNum ).SupFanNum = UnitarySystem( UnitarySysNum ).FanIndex;
+			PrimaryAirSystem( CurSysNum ).SupFanNum = UnitarySystem( UnitarySysNum ).FanIndex;
```
Indentation looks wrong here.

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -6018,19 +6018,40 @@ namespace HVACUnitarySystem {
-				ShowContinueError( "Occurs in " + CurrentModuleObject + " = " + UnitarySystem( UnitarySysNum ).Name );
+			if (errFlag) {
+				ShowContinueError("Occurs in " + CurrentModuleObject + " = " + UnitarySystem(UnitarySysNum).Name);
 				ErrorsFound = true;
```
Lost the space padding here.

rraustad replied:
Done

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -8538,7 +8559,7 @@ namespace HVACUnitarySystem {
 		} else if ( SELECT_CASE_var == CoilDX_MultiSpeedCooling ) { // Coil:Cooling:DX:Multispeed
 
-			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressor ) {
+			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressorCooling) {
```
And one here at the end and one or more below.

rraustad replied:
Done

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -9825,14 +9860,14 @@ namespace HVACUnitarySystem {
 										} else {
-											SolveRegulaFalsi( HumRatAcc, MaxIte, SolFla, SpeedRatio, DXCoilVarSpeedHumRatResidual, 0.0, 1.0, Par );
+										SolveRegulaFalsi( HumRatAcc, MaxIte, SolFla, SpeedRatio, DXCoilVarSpeedHumRatResidual, 0.0, 1.0, Par );
 										}
```
Indentation.

rraustad replied:
Done

mjwitte commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -9825,14 +9860,14 @@ namespace HVACUnitarySystem {
-										}
+										SpeedRatio = 1.0;
+									}
 									}
```
Indentation

rraustad replied:
Done

rraustad commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -8719,7 +8740,7 @@ namespace HVACUnitarySystem {
 		} else if ( SELECT_CASE_var == CoilDX_MultiSpeedHeating ) {
 
-			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressor ) {
+			if ( OutsideDryBulbTemp > UnitarySystem( UnitarySysNum ).MinOATCompressorHeating) {
```
spacing at end fixed

rraustad commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -9965,7 +10000,7 @@ namespace HVACUnitarySystem {
 			mdot = PartLoadFrac * UnitarySystem( UnitarySysNum ).MaxCoolCoilFluidFlow;
 			SetComponentFlowRate( mdot, UnitarySystem( UnitarySysNum ).CoolCoilFluidInletNode, UnitarySystem( UnitarySysNum ).CoolCoilFluidOutletNodeNum, UnitarySystem( UnitarySysNum ).CoolCoilLoopNum, UnitarySystem( UnitarySysNum ).CoolCoilLoopSide, UnitarySystem( UnitarySysNum ).CoolCoilBranchNum, UnitarySystem( UnitarySysNum ).CoolCoilCompNum );
-		}
+	}
```
fixed indentation

rraustad commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -10471,7 +10513,7 @@ namespace HVACUnitarySystem {
 			mdot = PartLoadFrac * UnitarySystem( UnitarySysNum ).MaxHeatCoilFluidFlow;
 			SetComponentFlowRate( mdot, UnitarySystem( UnitarySysNum ).HeatCoilFluidInletNode, UnitarySystem( UnitarySysNum ).HeatCoilFluidOutletNodeNum, UnitarySystem( UnitarySysNum ).HeatCoilLoopNum, UnitarySystem( UnitarySysNum ).HeatCoilLoopSide, UnitarySystem( UnitarySysNum ).HeatCoilBranchNum, UnitarySystem( UnitarySysNum ).HeatCoilCompNum );
-		}
+	}
```
fixed indentation

rraustad commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -10788,10 +10830,10 @@ namespace HVACUnitarySystem {
 	}
 
+		}
+
```
fixed indentation

rraustad committed:
fix indentation issues

rraustad committed:
Add min OAT for compressor operation to merged unit test

mjwitte commented:
So, those changes I suggested for the dx heating coil minimums have already been done in #6190.

rraustad commented:
@MyOldMopar I see these warnings that don't make sense. All I did was add 1 line to a unit test and now there are 2 unit tests that fail?

For i386-Windows-7-VisualStudio-14:

unknown file: error: C++ exception with description "invalid string position" thrown in the test body.

Myoldmopar commented:
Yeah, these sorts of anomalies are what @jasondegraw has been working on figuring out.  Even within Windows 32 and 64 bit, the results are different.  So. hard. to. debug.  I don't think you need to do anything on this PR to address those warnings.

Myoldmopar committed:
Add transition rules for 6158

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar commented:
Merged develop, which got the latest transition updates in place.  I then wrote the actual transition rules for this.  Built and sample tested, seemed to be OK.  I see there are many adjustments to the min-fields specifications, but just so you know, the transition program is currently writing the new field out at the end, as a blank, for (at least) `Coil:Cooling:DX:SingleSpeed` objects:

![image](https://user-images.githubusercontent.com/849824/28668906-9806f06e-7297-11e7-8899-1265c9d0a69e.png)

Is that what you are wanting?  Or do we need to refine some min fields/required field shenanigans? 

@mjwitte if there's anything I can do to help wrap this one up, let me know.

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -1230,13 +1231,17 @@ namespace DXCoils {
-			DXCoil( DXCoilNum ).LatentCapacityTimeConstant( 1 ) = Numbers( 9 );
+			//Set minimum OAT for compressor operation
+			DXCoil( DXCoilNum ).MinOATCompressor = Numbers( 6 );
+			if ( NumNumbers < 6 ) DXCoil( DXCoilNum ).MinOATCompressor = minOATCompDXCooling; // input field is after min fields and won't default if field not included 
```
@MyOldMopar @mjwitte the new input for min OAT comp in Coil:Cooing:DX:SingleSpeed is in field 15 when min-fields is 14. It seemed like the most logical place. I have this check in place to ensure MinOATCompressor is initialized to the existing default. If someone changes the default in the IDD and does not enter this field, it will still default to -25. If this field is entered and blank, the new user defined default min OAT will be used. Just as we know a field is blank, I would like to know the IDD field default so I can use that default in places like this.

Myoldmopar replied:
- So you have field A, with a default value in the IDD, but it can be overridden by a user.
- Then you have field B, and if the user leaves it blank, you want it to default to whatever the original default in the IDD was for field A?

rraustad replied:
No, I'm talking about the same field (field A in your example). Those fields that are past min fields have a problem with defaults in the IDD if the user does not enter this field (NumFields < this field number). The developer does not know what the default is if the user changes it and then doesn't enter that field as a blank in the object. I guess this is one reason we have min fields.

Instead of defaulting like this to -25C (hard coded), we could default to NumericFieldDefault(6) and then all defaults past min fields could be initialized in code and the user would still be able to change defaults in the IDD and not worry about entering a blank field (after min fields). Why should a user or GUI have to worry about these latter fields when there are defaults. MinFields should be just that, the very minimum.

Myoldmopar replied:
Yeah, this one is always tricky.  In my mind, I think of `\min-fields` as

> "the minimum number of fields that must be interpreted - blank or otherwise - by the simulation model in order to perform a simulation"

If you are saying you need a default to be inserted there by forcing it to be blank in order to make a good decision on your modeling, then I say it should be in the min fields.  Not sure this is the right time to have such a philosophical discussion about min fields, so for this PR just make it sufficient for a robust calculation with your new code and carry on.  Maybe we'll have those philosophical discussions over a hackathon sometime...

All that said, let me know what changes you are planning to make here and I'll wrap up my review when it is all ready to go.

rraustad commented on src/Transition/CreateNewIDFUsingRulesV8_8_0.f90:
```diff
@@ -365,6 +365,63 @@ SUBROUTINE CreateNewIDFUsingRules(EndOfFile,DiffOnly,InLfn,AskForInput,InputFile
+                 OutArgs(1:17)=InArgs(1:17)  ! No change
+                 OutArgs(18:CurArgs-1)=InArgs(19:CurArgs)  ! remove redundant field, shifting all fields up
+                 CurArgs = CurArgs - 1
+
```
I feel I should be making these changes to the transition program?

Myoldmopar replied:
Do you mean in general or for this PR?  If you mean in general then no.  Hopefully we won't have the transition program that much longer so don't invest the time.  If you mean for this PR then, well, I hope it's done...

rraustad commented:
I have no more changes here unless there are comments that need to be addressed.

rraustad commented:
602 files failing with in-bucket, get-bucket failing on Ubuntu. Sounds like what I'm thinking right now.

Myoldmopar commented:
> 602 files failing with in-bucket, get-bucket failing on Ubuntu. Sounds like what I'm thinking right now.

```
File "/usr/local/lib/python2.7/dist-packages/boto/s3/connection.py", line 539, in head_bucket
```

Failing in `head_bucket` makes me think of [Buckethead](https://en.wikipedia.org/wiki/Buckethead), but that's a **totally** different thing...

![image](https://user-images.githubusercontent.com/849824/28700598-64546c4a-7316-11e7-991e-104e0579d3c9.png)

rraustad commented:
KFC - Keep Frickin Crankin. I'm just getting tired of this pace and not sure how much longer I can keep it up. It's day and night for me and having it's toll.

rraustad commented:
Delete these last comments, they don't belong here.

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -9073,6 +9095,13 @@ namespace HVACUnitarySystem {
+			OutdoorDryBulb = Node( UnitarySystem( UnitarySysNum ).CondenserNodeNum ).Temp;
+		} else {
+			OutdoorDryBulb = OutDryBulbTemp;
+		}
```
Just a clarification for me.  `OutdoorDryBulb` should really be renamed to `CondenserInletAirTemp`, right?  Maybe it's a large change, so not in scope here, I'm just wanting to understand why this block wasn't there before.

rraustad replied:
This block was not there before because the model didn't check MinOAT to see if the compressor could operate. I added this at line 9137 to not bother with testing the output as if the compressor was on when the coil model would model it as off.

    // disable compressor if OAT is below minimum outdoor temperature
    if ( OutdoorDryBulb < UnitarySystem( UnitarySysNum ).MinOATCompressorCooling ) {
        SensibleLoad = false;
        LatentLoad = false;
    }

rraustad replied:
Oh, and because the cooling coil didn't have this before. And yes, it's the CondenserInletAirTemp.

Myoldmopar commented on src/EnergyPlus/PackagedTerminalHeatPump.cc:
```diff
@@ -6940,6 +6925,57 @@ namespace PackagedTerminalHeatPump {
+		std::string const HeatingCoilType, // type of heating coil
+		std::string const HeatingCoilName, // name of heating coil
+		bool & ErrorsFound // GetInput logical that errors were found
+	) {
```
This function looks like it is called once, so I am less concerned about the strings.  If any more changes are made we should go ahead and make them pass by reference.

rraustad replied:
Change to coil index is preferred (int CoilType, int CoilIndex).

rraustad replied:
@mjwitte also has some changes and suggested a new issue to resolve all comments noted here.
 
- unify use of CondenserInletAirTemp between parent and child (both use same information)
- change function call to use int's
- revise input field name? to Minimum Condenser Inlet Air Temperature for Compressor Operation
- others

Myoldmopar commented on src/ExpandObjects/epfilter.f90:
```diff
@@ -31850,6 +31855,7 @@ SUBROUTINE CreateDOAS
          base + doasNameOff,' DX Coil EIR-FF')
       CALL AddToObjFld('Part Load Fraction Correlation Curve Name',   &
          base + doasNameOff,' DX Coil PLF')
+      CALL AddToObjStr('Minimum Outdoor Dry-Bulb Temperature for Compressor Operation {C}','')
```
@mjwitte These all look right to me.  I'll build and run expand objects on a test file to verify.

Myoldmopar replied:
Built expand objects locally, ran transition and expand objects from this branch as well as the develop versions, diffing files along the way.  Everything looks great.

Myoldmopar commented on tst/EnergyPlus/unit/DXCoils.unit.cc:
```diff
@@ -1076,6 +1077,8 @@ namespace EnergyPlus {
 		EXPECT_EQ( 25000.0, DXCoil( 1 ).RatedTotCap( 1 ) );
 		EXPECT_EQ( DXCoil( 1 ).RatedTotCap( 1 ) * 0.004266, DXCoil( 1 ).EvapCondPumpElecNomPower( 1 ) );
+		// Minimum Outdoor Temperature for Compressor Operation defaults to -25.0 C
+		EXPECT_EQ( DXCoil( 1 ).MinOATCompressor, -25.0 );
```
Tagging onto an existing test; I like it.

Myoldmopar reviewed (changes requested):
@rraustad This is very close.  The only thing I want to resolve is the string arguments.  There's always this effort to squeeze performance out, so when there's string processing or a string argument involved, we've gotta tread carefully to avoid slipping.

In your comment, you confirmed that we should be able to pass indices around.  If that's true, then let's shoot for that.  If there's still a reason to pass strings, we'll at least get them passed as best as possible.

Myoldmopar commented:
Even though the Linux results files (diffs, etc.) didn't get pushed to the results buckets, the performance information for the runs was recorded, and shows a performance regression of 1%.  May not sound like much but we are trying to make efforts to squeeze a few percent here or there in the other direction all the time.  So I think it's worth some time at least make sure the function calls are being made properly.

@mbadams5 would changing the arguments from `std::string const` to `const std::string &` be sufficient to address "pretty much" any performance issues?  That's vague, and maybe depends on what the function is actually doing with the string inside, but I'd like to know your thoughts anyway.

Myoldmopar commented:
OK.  I can open a new issue that captures the little side pieces of this in order to get this moved in.  I'll make the issue and let you append items as needed.

Myoldmopar commented:
And there is now a merge conflict just due to the rules file.  I'll address that online here and then merge.

Myoldmopar committed:
Merge branch 'develop' into #6149-Add-low-temperature-shut-off-to-air-cooled-DX-cooling-coils
